### ISSUE TITLE
Counter-based PRNGs

### ISSUE THREAD

**Issue by mondus** [FLAMEGPU/FLAMEGPU2#362]
We could consider the use of a counter based PRNG to minimise the storage of seeds

https://dl.acm.org/doi/10.1145/2063384.2063405

Need to conisder / benchmark:

+ [ ] Statiistical properties
+ [ ] Memory requirements
+ [ ] Unform sampling performance
+ [ ] log sampling performance (if required)

**Comment by ptheywood** [comment]
Looks like curand includes a philox generator `CURAND_RNG_PHILOX4_32_10` (docs)[https://docs.nvidia.com/cuda/curand/host-api-overview.html#generator-types]

**Comment by Robadob** [comment]
I presume you've taken this idea from mrcgida's implementation?

**Comment by Robadob** [comment]
> From a random 2014 SO answer it looks like the curand philox actually has a larger state size than xorwow? Though the claim about running out of memory is bold.

Taken from my curand header.
```c
struct curandStatePhilox4_32_10 {
   uint4 ctr;                            // 16 bytes
   uint4 output;                         // 16 bytes
   uint2 key;                            // 8 bytes
   unsigned int STATE;                   // 4 bytes
   int boxmuller_flag;                   // 4 bytes
   int boxmuller_flag_double;            // 4 bytes
   float boxmuller_extra;                // 4 bytes
   double boxmuller_extra_double;        // 8 bytes
};                                       // 64 bytes total
struct curandStateXORWOW {
    unsigned int d, v[5];                // 6 x 4 bytes
    int boxmuller_flag;                  // 4 bytes
    int boxmuller_flag_double;           // 4 bytes
    float boxmuller_extra;               // 4 bytes
    double boxmuller_extra_double;       // 8 bytes
};                                       // 44 bytes total
```

It looks asthough it would be as simple as swapping the type from the default `curandState_t` typedef if we wanted to change.

**Comment by ptheywood** [comment]
Wrote a quick benchmarking app for the 3 easy to switch between pseudo generators: https://github.com/ptheywood/curand-benchmarking

Benchmarking on a Titan V, using 262k threads/states, each generating 65K uniformly distributed floats:


| engine | Initialisation (ms) | uniform fp32 (ms) |
|--------|---------------------|-------------------|
| XORWOW | 22.4766 | 20.2519 |
| MRG32k3a | 5.0403 | 103.4771 |
| Philox4_32_10 | 0.1398 | 68.757 |


 
![image](https://user-images.githubusercontent.com/628937/97328366-dd67c780-186d-11eb-8aac-c827fb50b797.png)


1070 results using the same number of threads showed much worse MRG performance, though 260k threads is a huge oversubscription, so probably better to modify the benchmark to launch a full device worth of threads with a grid stride loop.

**Comment by ptheywood** [comment]
After some discussion on slack w/ @Robadob, re curand init time looked into the performance of these out of curiosity, with recent CUDA's showing reduced XORWOW init time.

So I build and ran the benchmark above on my 3060ti using > 200k threads, and 65k random values per thread (which makes MRG look bad as it did on the 1070, so y limits are cropped) for CUDA 10.2 to 11.5

`CUDA >= 11.1` appears to have brought reasonable performance improvements to the init of `XORWOW`, but also the `uniformf` for `Philox4`

![3060ti](https://user-images.githubusercontent.com/628937/143610054-7f3f077c-4b87-4d3d-8c65-53d51db05d31.png)


(CSV and terrible plotting script included below incase they are needed again)

`3060ti.csv`
```
CUDA,engine,threads,samples_per_thread,repetitions,mean_bytes,mean_alloc_ms,mean_init_ms,mean_uniformf_ms,mean_dealloc_ms
10.2, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.2032,39.3020,32.3376,0.2751
10.2, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.2152,94.3408,3519.9778,0.1355
10.2, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.3318,0.1094,109.8158,0.2400
11.0, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.3868,29.1060,23.6411,0.1108
11.0, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.4815,101.7075,3665.4532,0.1983
11.0, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.4722,0.1155,95.9800,0.1085
11.1, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.2164,3.9431,27.5014,0.1329
11.1, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.4535,106.3768,3613.1803,0.1031
11.1, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.5683,0.0887,76.3974,0.1552
11.2, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.2164,3.9431,27.5014,0.1329
11.2, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.4535,106.3768,3613.1803,0.1031
11.2, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.5683,0.0887,76.3974,0.1552
11.3, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.1433,4.7368,36.5087,0.5536
11.3, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.1887,104.1201,3672.4001,0.1655
11.3, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.2543,0.1023,77.6554,0.1688
11.4, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.2030,4.6162,32.1045,0.2726
11.4, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.1981,94.7634,3400.4355,0.1894
11.4, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.3328,0.1024,73.3067,0.0767
11.5, 17curandStateXORWOW,262144,65536,5,12582912.000000,0.2626,3.8171,24.8936,0.0530
11.5, 19curandStateMRG32k3a,262144,65536,5,12582912.000000,0.3029,93.2815,3397.5163,0.0993
11.5, 24curandStatePhilox4_32_10,262144,65536,5,16777216.000000,0.2759,0.1021,69.9806,0.1255



```

`plot.py`
```
#! /usr/bin/python3

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt 

def main():
    CSV_IN = "3060ti.csv"
    OUTPUT_FILE = "3060ti.png"
    df = pd.read_csv(CSV_IN)
    sns.set_theme(style="whitegrid")
    fig, ax =plt.subplots(1,2, constrained_layout=True)
    sns.barplot(
        data=df,
        x="CUDA", y="mean_init_ms", hue="engine", palette="dark",
        ax=ax[0]
    )
    ax[0].title.set_text('Init')
    ax[1].title.set_text('uniformfloatf (65k)')
    sns.barplot(
        data=df,
        x="CUDA", y="mean_uniformf_ms", hue="engine", palette="dark",
        ax=ax[1]
    )
    ax[0].set_ylim([0, 40])
    ax[1].set_ylim([0, 400])

    handles, labels = ax[1].get_legend_handles_labels()
    bbox = (0.5, -0.1)
    legend = fig.legend(handles, labels, loc='lower center', ncol=1, bbox_to_anchor=bbox)

    ax[0].get_legend().set_visible(False)
    ax[1].get_legend().set_visible(False)

    plt.savefig(OUTPUT_FILE, bbox_inches='tight')
    plt.show()

if __name__ == "__main__":
    main()

```

**Comment by ptheywood** [comment]
As of CUDA 11.5.1, MRG323ka performance has supposidly been improved, although given we won't require >= 11.5.1 any time soon its' not going to be relevant to FLAME GPU 2. 


> Improved performance of CURAND_RNG_PSEUDO_MRG32K3A pseudorandom number generator when using ordering CURAND_ORDERING_PSEUDO_BEST or CURAND_ORDERING_PSEUDO_DEFAULT.

**Comment by ptheywood** [comment]
If the change to philox is made, it might also be worth exposing the `float4`/`double2` methods to users

> The above functions generate four single precision or two double precision results with each call. Because the underlying implementation uses the Philox generator, this is generally more efficient than generating a single result with each call.

**Comment by ptheywood** [comment]
Re-ran my 3060 ti bench from before with 11.6 and 11.7 to show the improvments to mrg use. 
Init has been improved too, but is still off the chart. (86s vs >100s pre 11.4 and ~ 94s pre 11.6)

![3060ti](https://user-images.githubusercontent.com/628937/174598588-4303a932-c60d-4245-ac2b-e0c12ff8c5d6.png)

I might look at adding float4 to this bench (and maybe normal dists) to this benching if I have half an hour spare at some point, to see if that shifts the balance towards philox as the default.

**Comment by ptheywood** [comment]
Closed as part of #873, which adds a CMake configuration option for controllwing which of 3 Curand implementations to use, defaulting to `PHILOX` which is a counter based prng.
